core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -336,15 +323,15 @@ public T acquire(long time, TimeUnit unit)
    */
   @Override
   public void release(T resource) {
+    if (!mResources.containsKey(resource)) {
+      throw new IllegalArgumentException(
+          "Resource " + resource.toString() + " was not acquired from this resource pool.");
+    }
+    ResourceInternal<T> resourceInternal = mResources.get(resource);
+    resourceInternal.setLastAccessTimeMs(mClock.millis());
     try {
       mLock.lock();
-      if (!mResources.containsKey(resource)) {
-        throw new IllegalArgumentException(
-            "Resource " + resource.toString() + " was not acquired from this resource pool.");
-      }
-      ResourceInternal<T> resourceInternal = mResources.get(resource);
-      resourceInternal.setLastAccessTimeMs(mClock.millis());
-      mResourceAvailable.add(resourceInternal);
+      mResourceAvailable.addFirst(resourceInternal);
```

Comment:
should we check `contains` again here after lock, to be thread-safe for a corner case of two concurrent `releases` releasing the same resource? It looks to me that this current code may result in adding back this resource twice.